internal/gps/pkgtree/digest.go

```diff
@@ -0,0 +1,442 @@
+		} else {
+			written, err = io.Copy(h, newLineEndingReader(fh))            // fast copy of file contents to hash
+			err = errors.Wrap(err, "cannot Copy")                         // errors.Wrap only wraps non-nil, so elide guard condition
+			writeBytesWithNull(h, []byte(strconv.FormatInt(written, 10))) // format file size as base 10 integer
+		}
+
+		// Close the file handle to the open directory without masking possible
+		// previous error value.
+		if er := fh.Close(); err == nil {
+			err = errors.Wrap(er, "cannot Close")
+		}
+		if err != nil {
+			return nil, err // early termination iff error
+		}
+	}
+
+	return h.Sum(nil), nil
+}
+
+// lineEndingReader is a `io.Reader` that converts CRLF sequences to LF.
+//
+// Some VCS systems automatically convert LF line endings to CRLF on some OS
+// platforms. This would cause the a file checked out on those platforms to have
+// a different digest than the same file on platforms that do not perform this
+// translation. In order to ensure file contents normalize and hash the same,
+// this struct satisfies the io.Reader interface by providing a Read method that
+// modifies the file's contents when it is read, translating all CRLF sequences
+// to LF.
+type lineEndingReader struct {
+	src             io.Reader // source io.Reader from which this reads
+	prevReadEndedCR bool      // used to track whether final byte of previous Read was CR
+}
+
+// newLineEndingReader returns a new lineEndingReader that reads from the
+// specified source io.Reader.
+func newLineEndingReader(src io.Reader) *lineEndingReader {
+	return &lineEndingReader{src: src}
+}
+
+// Read consumes bytes from the structure's source io.Reader to fill the
+// specified slice of bytes. It converts all CRLF byte sequences to LF, and
+// handles cases where CR and LF straddle across two Read operations.
+func (f *lineEndingReader) Read(buf []byte) (int, error) {
+	buflen := len(buf)
+	if f.prevReadEndedCR {
+		// Read one less byte in case we need to insert CR in there
+		buflen--
+	}
+	nr, er := f.src.Read(buf[:buflen])
+	if nr > 0 {
+		if f.prevReadEndedCR && buf[0] != '\n' {
+			// Having a CRLF split across two Read operations is rare, so
+			// ignoring performance impact of copying entire buffer by one
+			// byte. Plus, `copy` builtin likely uses machien opcode for
+			// performing the memory copy.
+			copy(buf[1:nr+1], buf[:nr]) // shift data to right one byte
+			buf[0] = '\r'               // insert the previous skipped CR byte at start of buf
+			nr++                        // pretend we read one more byte
+		}
+
+		// Remove any CRLF sequences in buf, using `bytes.Index` because that
+		// takes advantage of machine opcodes that search for byte patterns on
+		// many architectures.
+		var prevIndex int
+		for {
+			index := bytes.Index(buf[prevIndex:nr], []byte("\r\n"))
+			if index == -1 {
+				break
+			}
+			// Want to skip index byte, where the CR is.
+			copy(buf[prevIndex+index:nr-1], buf[prevIndex+index+1:nr])
```

Comment:
First of all, thanks for commenting on this issue.

The multiple copying here was somewhat plaguing to me as well. A couple of factors went into my decision. First, I wanted the algorithm to take advantage of the efficiencies of searching left-to-right in a memory sequence for a byte sequence, using the same opcodes that `man memmem` would call. I figured it would be faster than a right hand search. And I checked the `bytes.Index` implementation to see that it indeed does invoke some assembly to get the job done with the short target string we use.

Second, I am pretty sure that `builtin`'s `copy` takes advantage of the same machine opcodes for copying byte sequences that `man memmove` uses.

So, given those two reasons, I opted for that method, despite the nagging feeling there might be a better way. So, I'm glad you said something about it. Now, getting down in the details:

I wrongly thought that whether we started on the left and recopied things over and over, or started on the right, we'd still be recopying things over and over. The difference is how many things we are copying on subsequent copies.

Consider the following very short example, where `cd` is to be replaced by `d`:

`abcdeabcdeabcde`

Using the search-from-the-left method, and pulling everything on the right over one byte, and repeating, results in a total 23 bytes getting copied. Whereas, using the search-from-the-right method, and copy those bytes over, we only have 18 bytes copied. That's about a 20% improvement in a very short sequence of bytes. I'm sure for files which are tens of kilobytes long, the savings effect would be more pronounced.

After reading your algorithm, and a bit of work, I came up with a two loop system. The first loop would create a slice of indexes where the CRLF were located, and the second loop would copy substrings over to cover up the CR bytes, and no byte would be copied twice.

Then I created a single loop solution that runs through the byte slice, searching for the next CRLF sequence, and when it finds one, it shifts the entire substring from the previous CRLF sequence to this CRLF sequence back one byte. 👍